src/arrowc/schema.c

```diff
@@ -0,0 +1,236 @@
+        schema->dictionary->release(schema->dictionary);
+      }
+
+      ARROWC_FREE(schema->dictionary);
+    }
+
+    // private data not currently used
+    if (schema->private_data != NULL) {
+      ARROWC_FREE(schema->private_data);
+    }
+
+    schema->release = NULL;
+  }
+}
+
+int ArrowSchemaInit(int64_t n_children, struct ArrowSchema* schema) {
+  schema->format = NULL;
+  schema->name = NULL;
+  schema->metadata = NULL;
+  schema->flags = ARROW_FLAG_NULLABLE;
+  schema->n_children = n_children;
+  schema->children = NULL;
+  schema->dictionary = NULL;
+  schema->private_data = NULL;
+  schema->release = &ArrowSchemaRelease;
+
+  if (n_children > 0) {
+    schema->children =
+        (struct ArrowSchema**)ARROWC_MALLOC(n_children * sizeof(struct ArrowSchema*));
+
+    if (schema->children == NULL) {
+      schema->release(schema);
+      return ENOMEM;
+    }
+
+    memset(schema->children, 0, n_children * sizeof(struct ArrowSchema*));
+
+    for (int64_t i = 0; i < n_children; i++) {
+      schema->children[i] =
+          (struct ArrowSchema*)ARROWC_MALLOC(sizeof(struct ArrowSchema));
+
+      if (schema->children[i] == NULL) {
+        schema->release(schema);
+        return ENOMEM;
+      }
+
+      schema->children[i]->release = NULL;
+    }
+  }
+
+  // We don't allocate the dictionary because it has to be nullptr
+  // for non-dictionary-encoded arrays.
+  return ARROWC_OK;
+}
+
+static inline int64_t ArrowSchemaMetadataSize(const char* metadata) {
+  if (metadata == NULL) {
+    return 0;
+  }
+
+  int64_t pos = 0;
+  int32_t n;
+  memcpy(&n, metadata + pos, sizeof(int32_t));
+  pos += sizeof(int32_t);
+
+  for (int i = 0; i < n; i++) {
+    int32_t name_size;
+    memcpy(&name_size, metadata + pos, sizeof(int32_t));
+    pos += sizeof(int32_t);
+
+    if (name_size > 0) {
+      pos += name_size;
+    }
+
+    int32_t value_size;
+    memcpy(&value_size, metadata + pos, sizeof(int32_t));
+    pos += sizeof(int32_t);
+
+    if (value_size > 0) {
+      pos += value_size;
+    }
+  }
+
+  return pos;
+}
+
+ArrowErrorCode ArrowSchemaSetFormat(struct ArrowSchema* schema, const char* format) {
+  if (format != NULL) {
+    size_t format_size = strlen(format) + 1;
+    schema->format = (const char*)ARROWC_MALLOC(format_size);
+    if (schema->format == NULL) {
+      return ENOMEM;
+    }
+
+    memcpy((void*)schema->format, format, format_size);
+  } else {
+    schema->format = NULL;
+  }
+
+  return ARROWC_OK;
+}
+
+ArrowErrorCode ArrowSchemaSetName(struct ArrowSchema* schema, const char* name) {
+  if (name != NULL) {
+    size_t name_size = strlen(name) + 1;
+    schema->name = (const char*)ARROWC_MALLOC(name_size);
+    if (schema->name == NULL) {
+      return ENOMEM;
+    }
+
+    memcpy((void*)schema->name, name, name_size);
+  } else {
+    schema->name = NULL;
+  }
+
+  return ARROWC_OK;
+}
+
+ArrowErrorCode ArrowSchemaSetMetadata(struct ArrowSchema* schema, const char* metadata) {
```

Comment:
This API seems a bit weird. Wouldn't callers instead want to build metadata by passing key-value pairs? For now it seems they have to the encoding themselves in their own buffer, _and_ you then copy the buffer after manually recomputing the metadata size (which the caller presumably already knows since it has just built it up).